// @pulumi/aws-compliance-policies/awsnative/apigateway/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.DomainName = void 0;
const apigateway_1 = require("@pulumi/aws-native/apigateway");
const policy_1 = require("@pulumi/policy");
const compliance_policy_manager_1 = require("@pulumi/compliance-policy-manager");
var DomainName;
(function (DomainName) {
    /**
     * Checks that ApiGateway Domain Name Security Policy uses secure/modern TLS encryption.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, network
     * @link https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-custom-domain-tls-version.html
     */
    DomainName.configureSecurityPolicy = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "awsnative-apigateway-domainname-configure-security-policy",
            description: "Checks that ApiGateway Domain Name Security Policy uses secure/modern TLS encryption.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(apigateway_1.DomainName, (domainName, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!domainName.securityPolicy || domainName.securityPolicy !== "TLS_1_2") {
                    reportViolation("API Gateway Domain Name Security Policy should use secure/modern TLS encryption.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["apigateway"],
        severity: "high",
        topics: ["network", "encryption"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
})(DomainName || (DomainName = {}));
exports.DomainName = DomainName;
