// @pulumi/aws-compliance-policies/aws/efs/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSystem = void 0;
const efs_1 = require("@pulumi/aws/efs");
const policy_1 = require("@pulumi/policy");
const compliance_policy_manager_1 = require("@pulumi/compliance-policy-manager");
var FileSystem;
(function (FileSystem) {
    /**
     * Check that encrypted EFS File system uses a customer-managed KMS key.
     *
     * @severity low
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, storage
     * @link https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html
     */
    FileSystem.configureCustomerManagedKey = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-efs-filesystem-configure-customer-managed-key",
            description: "Check that encrypted EFS File system uses a customer-managed KMS key.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(efs_1.FileSystem, (fileSystem, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (fileSystem.encrypted && !fileSystem.kmsKeyId) {
                    reportViolation("An EFS File System should be encrypted using a customer-managed KMS key.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["efs"],
        severity: "low",
        topics: ["encryption", "storage"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Check that EFS File system doesn't use single availability zone.
     *
     * @severity high
     * @frameworks none
     * @topics availability, storage
     * @link https://docs.aws.amazon.com/efs/latest/ug/storage-classes.html
     */
    FileSystem.disallowSingleAvailabilityZone = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-efs-filesystem-disallow-single-availability-zone",
            description: "Check that EFS File system doesn't use single availability zone.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(efs_1.FileSystem, (fileSystem, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (fileSystem.availabilityZoneName) {
                    reportViolation("EFS File Systems should use more than one availability zone.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["efs"],
        severity: "high",
        topics: ["storage", "availability"],
    });
    /**
     * Checks that EFS File Systems do not have an unencrypted file system.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, storage
     * @link https://docs.aws.amazon.com/efs/latest/ug/encryption-at-rest.html
     */
    FileSystem.disallowUnencryptedFileSystem = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-efs-filesystem-disallow-unencrypted-file-system",
            description: "Checks that EFS File Systems do not have an unencrypted file system.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(efs_1.FileSystem, (fileSystem, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!fileSystem.encrypted) {
                    reportViolation("EFS File systems should not have an unencypted file system.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["efs"],
        severity: "high",
        topics: ["encryption", "storage"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
})(FileSystem || (FileSystem = {}));
exports.FileSystem = FileSystem;
