// @pulumi/aws-compliance-policies/awsnative/cloudfront/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.Distribution = void 0;
const cloudfront_1 = require("@pulumi/aws-native/cloudfront");
const policy_1 = require("@pulumi/policy");
const compliance_policy_manager_1 = require("@pulumi/compliance-policy-manager");
var Distribution;
(function (Distribution) {
    /**
     * Checks that any CloudFront distributions have access logging configured.
     *
     * @severity medium
     * @frameworks hitrust, iso27001, pcidss
     * @topics logging, network
     * @link https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/AccessLogs.html
     */
    Distribution.configureAccessLogging = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "awsnative-cloudfront-distribution-configure-access-logging",
            description: "Checks that any CloudFront distributions have access logging configured.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(cloudfront_1.Distribution, (distribution, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (distribution.distributionConfig.logging && !distribution.distributionConfig.logging.bucket) {
                    reportViolation("CloudFront Distributions should have access logging configured.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["cloudfront"],
        severity: "medium",
        topics: ["network", "logging"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that CloudFront distributions uses secure/modern TLS encryption.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, network
     * @link https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/secure-connections-supported-viewer-protocols-ciphers.html
     */
    Distribution.configureSecureTls = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "awsnative-cloudfront-distribution-configure-secure-tls",
            description: "Checks that CloudFront distributions uses secure/modern TLS encryption.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(cloudfront_1.Distribution, (distribution, args, reportViolation) => {
                var _a;
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (distribution.distributionConfig.viewerCertificate && ((_a = distribution.distributionConfig.viewerCertificate.minimumProtocolVersion) === null || _a === void 0 ? void 0 : _a.toLowerCase()) !== "TLSv1.2_2021".toLowerCase()) {
                    reportViolation("CloudFront distributions should use secure/modern TLS encryption.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["cloudfront"],
        severity: "high",
        topics: ["network", "encryption"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that CloudFront distributions communicate with custom origins using TLS 1.2 encryption only.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, network
     * @link https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-https-cloudfront-to-s3-origin.html
     */
    Distribution.configureSecureTlsToOrigin = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "awsnative-cloudfront-distribution-configure-secure-tls-to-origin",
            description: "Checks that CloudFront distributions communicate with custom origins using TLS 1.2 encryption only.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(cloudfront_1.Distribution, (distribution, args, reportViolation) => {
                var _a;
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                (_a = distribution.distributionConfig.origins) === null || _a === void 0 ? void 0 : _a.forEach((origin) => {
                    var _a, _b;
                    (_b = (_a = origin.customOriginConfig) === null || _a === void 0 ? void 0 : _a.originSslProtocols) === null || _b === void 0 ? void 0 : _b.forEach((sslProtocol) => {
                        if (sslProtocol.toLowerCase() !== "TLSv1.2".toLowerCase()) {
                            reportViolation("CloudFront Distributions should only use TLS 1.2 encryption to communicate with custom origins.");
                        }
                    });
                });
            }),
        },
        vendors: ["aws"],
        services: ["cloudfront"],
        severity: "high",
        topics: ["network", "encryption"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that CloudFront distributions have a WAF ACL associated.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics network
     * @link https://docs.aws.amazon.com/waf/latest/developerguide/cloudfront-features.html
     */
    Distribution.configureWafAcl = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "awsnative-cloudfront-distribution-configure-waf-acl",
            description: "Checks that CloudFront distributions have a WAF ACL associated.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(cloudfront_1.Distribution, (distribution, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!distribution.distributionConfig.webAclId) {
                    reportViolation("CloudFront Distributions should have a WAF ACL associated.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["cloudfront"],
        severity: "high",
        topics: ["network"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that CloudFront distributions only allow encypted ingress traffic.
     *
     * @severity critical
     * @frameworks hitrust, iso27001, pcidss
     * @topics network
     * @link https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol
     */
    Distribution.disallowUnencryptedTraffic = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "awsnative-cloudfront-distribution-disallow-unencrypted-traffic",
            description: "Checks that CloudFront distributions only allow encypted ingress traffic.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(cloudfront_1.Distribution, (distribution, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (distribution.distributionConfig.defaultCacheBehavior.viewerProtocolPolicy.toLowerCase() === "allow-all") {
                    reportViolation("CloudFront distributions should not allow unencrypted traffic.");
                }
                if (distribution.distributionConfig.cacheBehaviors) {
                    if (distribution.distributionConfig.cacheBehaviors.some((cacheBehaviors) => cacheBehaviors.viewerProtocolPolicy.toLowerCase() === "allow-all")) {
                        reportViolation("CloudFront distributions should not allow unencrypted traffic.");
                    }
                }
            }),
        },
        vendors: ["aws"],
        services: ["cloudfront"],
        severity: "critical",
        topics: ["network"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that any CloudFront distributions have access logging enabled.
     *
     * @severity medium
     * @frameworks hitrust, iso27001, pcidss
     * @topics logging, network
     * @link https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/AccessLogs.html
     */
    Distribution.enableAccessLogging = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "awsnative-cloudfront-distribution-enable-access-logging",
            description: "Checks that any CloudFront distributions have access logging enabled.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(cloudfront_1.Distribution, (distribution, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!distribution.distributionConfig.logging) {
                    reportViolation("CloudFront Distributions should have logging enabled.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["cloudfront"],
        severity: "medium",
        topics: ["network", "logging"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that CloudFront distributions communicate with custom origins using TLS encryption.
     *
     * @severity critical
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, network
     * @link https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-https-cloudfront-to-s3-origin.html
     */
    Distribution.enableTlsToOrigin = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "awsnative-cloudfront-distribution-enable-tls-to-origin",
            description: "Checks that CloudFront distributions communicate with custom origins using TLS encryption.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(cloudfront_1.Distribution, (distribution, args, reportViolation) => {
                var _a;
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                (_a = distribution.distributionConfig.origins) === null || _a === void 0 ? void 0 : _a.forEach((origin) => {
                    if (origin.customOriginConfig && (!origin.customOriginConfig.originProtocolPolicy || origin.customOriginConfig.originProtocolPolicy.toLowerCase() !== "https-only")) {
                        reportViolation("CloudFront Distributions should use TLS encryption to communicate with custom origins.");
                    }
                });
            }),
        },
        vendors: ["aws"],
        services: ["cloudfront"],
        severity: "critical",
        topics: ["network", "encryption"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
})(Distribution || (Distribution = {}));
exports.Distribution = Distribution;
