// @pulumi/aws-compliance-policies/aws/s3/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bucket = void 0;
const s3_1 = require("@pulumi/aws/s3");
const policy_1 = require("@pulumi/policy");
const compliance_policy_manager_1 = require("@pulumi/compliance-policy-manager");
var Bucket;
(function (Bucket) {
    /**
     * Checks that S3 Bucket have cross-region replication configured.
     *
     * @severity high
     * @frameworks iso27001, pcidss
     * @topics availability
     * @link https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication.html
     */
    Bucket.configureReplicationConfiguration = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-s3-bucket-configure-replication-configuration",
            description: "Checks that S3 Bucket have cross-region replication configured.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(s3_1.Bucket, (bucket, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (bucket.replicationConfiguration && bucket.replicationConfiguration.rules) {
                    bucket.replicationConfiguration.rules.forEach((rule) => {
                        if (rule.status.toLowerCase() !== "enabled") {
                            reportViolation("S3 Buckets replication rules should be configured.");
                        }
                    });
                }
            }),
        },
        vendors: ["aws"],
        services: ["s3"],
        severity: "high",
        topics: ["availability"],
        frameworks: ["pcidss", "iso27001"],
    });
    /**
     * Check that S3 Buckets Server-Side Encryption (SSE) is using a customer-managed KMS Key.
     *
     * @severity low
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, storage
     * @link https://docs.aws.amazon.com/AmazonS3/latest/userguide/specifying-kms-encryption.html
     */
    Bucket.configureServerSideEncryptionCustomerManagedKey = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-s3-bucket-configure-server-side-encryption-customer-managed-key",
            description: "Check that S3 Buckets Server-Side Encryption (SSE) is using a customer-managed KMS Key.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(s3_1.Bucket, (bucket, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (bucket.serverSideEncryptionConfiguration && !bucket.serverSideEncryptionConfiguration.rule.applyServerSideEncryptionByDefault.kmsMasterKeyId) {
                    reportViolation("S3 Buckets Server-Side Encryption (SSE) should use a Customer-managed KMS key.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["s3"],
        severity: "low",
        topics: ["encryption", "storage"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Check that S3 Buckets Server-Side Encryption (SSE) uses AWS KMS.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, storage
     * @link https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingKMSEncryption.html
     */
    Bucket.configureServerSideEncryptionKms = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-s3-bucket-configure-server-side-encryption-kms",
            description: "Check that S3 Buckets Server-Side Encryption (SSE) uses AWS KMS.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(s3_1.Bucket, (bucket, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (bucket.serverSideEncryptionConfiguration && bucket.serverSideEncryptionConfiguration.rule.applyServerSideEncryptionByDefault.sseAlgorithm.toLowerCase() !== "aws:kms") {
                    reportViolation("S3 Buckets Server-Side Encryption (SSE) should use AWS KMS.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["s3"],
        severity: "high",
        topics: ["encryption", "storage"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that S3 Bucket ACLs don't allow 'public-read' or 'public-read-write' or 'authenticated-read'.
     *
     * @severity critical
     * @frameworks cis, hitrust, iso27001, pcidss
     * @topics security, storage
     * @link https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-control-block-public-access.html
     */
    Bucket.disallowPublicRead = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-s3-bucket-disallow-public-read",
            description: "Checks that S3 Bucket ACLs don't allow 'public-read' or 'public-read-write' or 'authenticated-read'.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(s3_1.Bucket, (bucket, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (bucket.acl) {
                    if (bucket.acl.toLowerCase() === "public-read" || bucket.acl.toLowerCase() === "public-read-write" || bucket.acl.toLowerCase() === "authenticated-read") {
                        reportViolation("S3 Buckets ACLs should not be set to 'public-read', 'public-read-write' or 'authenticated-read'.");
                    }
                }
            }),
        },
        vendors: ["aws"],
        services: ["s3"],
        severity: "critical",
        topics: ["storage", "security"],
        frameworks: ["cis", "pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that S3 Bucket have cross-region replication enabled.
     *
     * @severity high
     * @frameworks iso27001, pcidss
     * @topics availability
     * @link https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication.html
     */
    Bucket.enableReplicationConfiguration = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-s3-bucket-enable-replication-configuration",
            description: "Checks that S3 Bucket have cross-region replication enabled.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(s3_1.Bucket, (bucket, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!bucket.replicationConfiguration || !bucket.replicationConfiguration.rules || bucket.replicationConfiguration.rules.length < 1) {
                    reportViolation("S3 buckets should have cross-region replication enabled.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["s3"],
        severity: "high",
        topics: ["availability"],
        frameworks: ["pcidss", "iso27001"],
    });
    /**
     * Check that S3 Bucket Server-Side Encryption (SSE) is enabled.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, storage
     * @link https://docs.aws.amazon.com/AmazonS3/latest/userguide/serv-side-encryption.html
     */
    Bucket.enableServerSideEncryption = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-s3-bucket-enable-server-side-encryption",
            description: "Check that S3 Bucket Server-Side Encryption (SSE) is enabled.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(s3_1.Bucket, (bucket, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!bucket.serverSideEncryptionConfiguration) {
                    reportViolation("S3 Buckets Server-Side Encryption (SSE) should be enabled.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["s3"],
        severity: "high",
        topics: ["encryption", "storage"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Check that S3 Buckets Server-Side Encryption (SSE) is using a Bucket key.
     *
     * @severity medium
     * @frameworks iso27001, pcidss
     * @topics cost, encryption, storage
     * @link https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-key.html
     */
    Bucket.enableServerSideEncryptionBucketKey = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-s3-bucket-enable-server-side-encryption-bucket-key",
            description: "Check that S3 Buckets Server-Side Encryption (SSE) is using a Bucket key.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(s3_1.Bucket, (bucket, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (bucket.serverSideEncryptionConfiguration && bucket.serverSideEncryptionConfiguration.rule.bucketKeyEnabled !== true) {
                    reportViolation("S3 Buckets Server-Side Encryption (SSE) should use a Bucket key to reduce cost.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["s3"],
        severity: "medium",
        topics: ["encryption", "storage", "cost"],
        frameworks: ["pcidss", "iso27001"],
    });
})(Bucket || (Bucket = {}));
exports.Bucket = Bucket;
