// @pulumi/aws-compliance-policies/aws/ebs/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.Volume = void 0;
const ebs_1 = require("@pulumi/aws/ebs");
const policy_1 = require("@pulumi/policy");
const compliance_policy_manager_1 = require("@pulumi/compliance-policy-manager");
var Volume;
(function (Volume) {
    /**
     * Check that encrypted EBS volumes use a customer-managed KMS key.
     *
     * @severity low
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, storage
     * @link https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html
     */
    Volume.configureCustomerManagedKey = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-ebs-volume-configure-customer-managed-key",
            description: "Check that encrypted EBS volumes use a customer-managed KMS key.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(ebs_1.Volume, (v, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (v.encrypted && !v.kmsKeyId) {
                    reportViolation("An EBS volume should be encrypted using a customer-managed KMS key.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["ebs"],
        severity: "low",
        topics: ["encryption", "storage"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that EBS volumes are encrypted.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, storage
     * @link https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html
     */
    Volume.disallowUnencryptedVolume = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-ebs-volume-disallow-unencrypted-volume",
            description: "Checks that EBS volumes are encrypted.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(ebs_1.Volume, (v, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!v.encrypted) {
                    reportViolation("An EBS volume is currently not encrypted.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["ebs"],
        severity: "high",
        topics: ["encryption", "storage"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
})(Volume || (Volume = {}));
exports.Volume = Volume;
