// @pulumi/aws-compliance-policies/aws/apigatewayv2/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stage = exports.DomainName = void 0;
const apigatewayv2_1 = require("@pulumi/aws/apigatewayv2");
const policy_1 = require("@pulumi/policy");
const compliance_policy_manager_1 = require("@pulumi/compliance-policy-manager");
const apigatewayv2_2 = require("@pulumi/aws/apigatewayv2");
var DomainName;
(function (DomainName) {
    /**
     * Checks that any ApiGatewayV2 Domain Name Security Policy uses secure/modern TLS encryption.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, network
     * @link https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-custom-domain-tls-version.html
     */
    DomainName.configureDomainNameSecurityPolicy = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-apigatewayv2-domainname-configure-domain-name-security-policy",
            description: "Checks that any ApiGatewayV2 Domain Name Security Policy uses secure/modern TLS encryption.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(apigatewayv2_1.DomainName, (domainName, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (domainName.domainNameConfiguration && domainName.domainNameConfiguration.securityPolicy !== "TLS_1_2") {
                    reportViolation("API GatewayV2 Domain Name Security Policy should use secure/modern TLS encryption.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["apigatewayv2"],
        severity: "high",
        topics: ["network", "encryption"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that any ApiGatewayV2 Domain Name Configuration is enabled.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics network
     * @link https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-custom-domain-tls-version.html
     */
    DomainName.enableDomainNameConfiguration = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-apigatewayv2-domainname-enable-domain-name-configuration",
            description: "Checks that any ApiGatewayV2 Domain Name Configuration is enabled.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(apigatewayv2_1.DomainName, (domainName, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!domainName.domainNameConfiguration) {
                    reportViolation("API GatewayV2 Domain Name Configuration should be enabled.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["apigatewayv2"],
        severity: "high",
        topics: ["network"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
})(DomainName || (DomainName = {}));
exports.DomainName = DomainName;
var Stage;
(function (Stage) {
    /**
     * Checks that any ApiGatewayV2 Stages have access logging configured.
     *
     * @severity medium
     * @frameworks hitrust, iso27001, pcidss
     * @topics logging, network
     * @link https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-logging.html
     */
    Stage.configureAccessLogging = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-apigatewayv2-stage-configure-access-logging",
            description: "Checks that any ApiGatewayV2 Stages have access logging configured.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(apigatewayv2_2.Stage, (stage, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (stage.accessLogSettings && (!stage.accessLogSettings.destinationArn || !stage.accessLogSettings.format)) {
                    reportViolation("API Gateway V2 stages should have access logging configured.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["apigatewayv2"],
        severity: "medium",
        topics: ["network", "logging"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Checks that any ApiGatewayV2 Stages have access logging enabled.
     *
     * @severity medium
     * @frameworks hitrust, iso27001, pcidss
     * @topics logging, network
     * @link https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-logging.html
     */
    Stage.enableAccessLogging = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-apigatewayv2-stage-enable-access-logging",
            description: "Checks that any ApiGatewayV2 Stages have access logging enabled.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(apigatewayv2_2.Stage, (stage, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!stage.accessLogSettings) {
                    reportViolation("API Gateway V2 stages should have access logging enabled.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["apigatewayv2"],
        severity: "medium",
        topics: ["network", "logging"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
})(Stage || (Stage = {}));
exports.Stage = Stage;
