// @pulumi/aws-compliance-policies/awsnative/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.s3 = exports.rds = exports.lambda = exports.kms = exports.eks = exports.efs = exports.ecr = exports.ec2 = exports.cloudfront = exports.athena = exports.appflow = exports.apigateway = void 0;
exports.apigateway = require("./apigateway");
exports.appflow = require("./appflow");
exports.athena = require("./athena");
exports.cloudfront = require("./cloudfront");
exports.ec2 = require("./ec2");
exports.ecr = require("./ecr");
exports.efs = require("./efs");
exports.eks = require("./eks");
exports.kms = require("./kms");
exports.lambda = require("./lambda");
exports.rds = require("./rds");
exports.s3 = require("./s3");
