// @pulumi/eks/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-gen-eks. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.types = exports.VpcCniAddon = exports.Provider = exports.NodeGroupV2 = exports.NodeGroupSecurityGroup = exports.NodeGroup = exports.ManagedNodeGroup = exports.ClusterCreationRoleProvider = exports.Addon = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
exports.Addon = null;
utilities.lazyLoad(exports, ["Addon"], () => require("./addon"));
__exportStar(require("./cluster"), exports);
const cluster_1 = require("./cluster");
exports.ClusterCreationRoleProvider = null;
utilities.lazyLoad(exports, ["ClusterCreationRoleProvider"], () => require("./clusterCreationRoleProvider"));
__exportStar(require("./clusterMixins"), exports);
exports.ManagedNodeGroup = null;
utilities.lazyLoad(exports, ["ManagedNodeGroup"], () => require("./managedNodeGroup"));
exports.NodeGroup = null;
utilities.lazyLoad(exports, ["NodeGroup"], () => require("./nodeGroup"));
exports.NodeGroupSecurityGroup = null;
utilities.lazyLoad(exports, ["NodeGroupSecurityGroup"], () => require("./nodeGroupSecurityGroup"));
exports.NodeGroupV2 = null;
utilities.lazyLoad(exports, ["NodeGroupV2"], () => require("./nodeGroupV2"));
__exportStar(require("./nodegroupMixins"), exports);
exports.Provider = null;
utilities.lazyLoad(exports, ["Provider"], () => require("./provider"));
__exportStar(require("./storageclassMixins"), exports);
exports.VpcCniAddon = null;
utilities.lazyLoad(exports, ["VpcCniAddon"], () => require("./vpcCniAddon"));
// Export enums:
__exportStar(require("./types/enums"), exports);
// Export sub-modules:
const types = require("./types");
exports.types = types;
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "eks:index:Addon":
                return new exports.Addon(name, undefined, { urn });
            case "eks:index:Cluster":
                return new cluster_1.Cluster(name, undefined, { urn });
            case "eks:index:ClusterCreationRoleProvider":
                return new exports.ClusterCreationRoleProvider(name, undefined, { urn });
            case "eks:index:ManagedNodeGroup":
                return new exports.ManagedNodeGroup(name, undefined, { urn });
            case "eks:index:NodeGroup":
                return new exports.NodeGroup(name, undefined, { urn });
            case "eks:index:NodeGroupSecurityGroup":
                return new exports.NodeGroupSecurityGroup(name, undefined, { urn });
            case "eks:index:NodeGroupV2":
                return new exports.NodeGroupV2(name, undefined, { urn });
            case "eks:index:VpcCniAddon":
                return new exports.VpcCniAddon(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("eks", "index", _module);
pulumi.runtime.registerResourcePackage("eks", {
    version: utilities.getVersion(),
    constructProvider: (name, type, urn) => {
        if (type !== "pulumi:providers:eks") {
            throw new Error(`unknown provider type ${type}`);
        }
        return new exports.Provider(name, undefined, { urn });
    },
});
