// @pulumi/eks/nodeGroupSecurityGroup.js
"use strict";
// *** WARNING: this file was generated by pulumi-gen-eks. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeGroupSecurityGroup = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * NodeGroupSecurityGroup is a component that wraps creating a security group for node groups with the default ingress & egress rules required to connect and work with the EKS cluster security group.
 */
class NodeGroupSecurityGroup extends pulumi.ComponentResource {
    /**
     * Returns true if the given object is an instance of NodeGroupSecurityGroup.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === NodeGroupSecurityGroup.__pulumiType;
    }
    /**
     * Create a NodeGroupSecurityGroup resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.clusterSecurityGroup === undefined) && !opts.urn) {
                throw new Error("Missing required property 'clusterSecurityGroup'");
            }
            if ((!args || args.eksCluster === undefined) && !opts.urn) {
                throw new Error("Missing required property 'eksCluster'");
            }
            if ((!args || args.vpcId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'vpcId'");
            }
            resourceInputs["clusterSecurityGroup"] = args ? args.clusterSecurityGroup : undefined;
            resourceInputs["eksCluster"] = args ? args.eksCluster : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["vpcId"] = args ? args.vpcId : undefined;
            resourceInputs["securityGroup"] = undefined /*out*/;
            resourceInputs["securityGroupRule"] = undefined /*out*/;
        }
        else {
            resourceInputs["securityGroup"] = undefined /*out*/;
            resourceInputs["securityGroupRule"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(NodeGroupSecurityGroup.__pulumiType, name, resourceInputs, opts, true /*remote*/);
    }
}
exports.NodeGroupSecurityGroup = NodeGroupSecurityGroup;
/** @internal */
NodeGroupSecurityGroup.__pulumiType = 'eks:index:NodeGroupSecurityGroup';
