// @pulumi/eks/nodeGroup.js
"use strict";
// *** WARNING: this file was generated by pulumi-gen-eks. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeGroup = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * NodeGroup is a component that wraps the AWS EC2 instances that provide compute capacity for an EKS cluster.
 *
 * @deprecated NodeGroup uses AWS EC2 LaunchConfiguration which has been deprecated by AWS and doesn't support the newest instance types. Please use NodeGroupV2 instead.
 */
class NodeGroup extends pulumi.ComponentResource {
    /**
     * Returns true if the given object is an instance of NodeGroup.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === NodeGroup.__pulumiType;
    }
    /**
     * Create a NodeGroup resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    /** @deprecated NodeGroup uses AWS EC2 LaunchConfiguration which has been deprecated by AWS and doesn't support the newest instance types. Please use NodeGroupV2 instead. */
    constructor(name, args, opts) {
        pulumi.log.warn("NodeGroup is deprecated: NodeGroup uses AWS EC2 LaunchConfiguration which has been deprecated by AWS and doesn't support the newest instance types. Please use NodeGroupV2 instead.");
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.cluster === undefined) && !opts.urn) {
                throw new Error("Missing required property 'cluster'");
            }
            resourceInputs["amiId"] = args ? args.amiId : undefined;
            resourceInputs["amiType"] = args ? args.amiType : undefined;
            resourceInputs["autoScalingGroupTags"] = args ? args.autoScalingGroupTags : undefined;
            resourceInputs["bootstrapExtraArgs"] = args ? args.bootstrapExtraArgs : undefined;
            resourceInputs["bottlerocketSettings"] = args ? args.bottlerocketSettings : undefined;
            resourceInputs["cloudFormationTags"] = args ? args.cloudFormationTags : undefined;
            resourceInputs["cluster"] = args ? args.cluster : undefined;
            resourceInputs["clusterIngressRule"] = args ? args.clusterIngressRule : undefined;
            resourceInputs["clusterIngressRuleId"] = args ? args.clusterIngressRuleId : undefined;
            resourceInputs["desiredCapacity"] = args ? args.desiredCapacity : undefined;
            resourceInputs["enableDetailedMonitoring"] = args ? args.enableDetailedMonitoring : undefined;
            resourceInputs["encryptRootBlockDevice"] = args ? args.encryptRootBlockDevice : undefined;
            resourceInputs["extraNodeSecurityGroups"] = args ? args.extraNodeSecurityGroups : undefined;
            resourceInputs["gpu"] = args ? args.gpu : undefined;
            resourceInputs["instanceProfile"] = args ? args.instanceProfile : undefined;
            resourceInputs["instanceProfileName"] = args ? args.instanceProfileName : undefined;
            resourceInputs["instanceType"] = args ? args.instanceType : undefined;
            resourceInputs["keyName"] = args ? args.keyName : undefined;
            resourceInputs["kubeletExtraArgs"] = args ? args.kubeletExtraArgs : undefined;
            resourceInputs["labels"] = args ? args.labels : undefined;
            resourceInputs["maxSize"] = args ? args.maxSize : undefined;
            resourceInputs["minSize"] = args ? args.minSize : undefined;
            resourceInputs["nodeAssociatePublicIpAddress"] = args ? args.nodeAssociatePublicIpAddress : undefined;
            resourceInputs["nodePublicKey"] = args ? args.nodePublicKey : undefined;
            resourceInputs["nodeRootVolumeDeleteOnTermination"] = args ? args.nodeRootVolumeDeleteOnTermination : undefined;
            resourceInputs["nodeRootVolumeEncrypted"] = args ? args.nodeRootVolumeEncrypted : undefined;
            resourceInputs["nodeRootVolumeIops"] = args ? args.nodeRootVolumeIops : undefined;
            resourceInputs["nodeRootVolumeSize"] = args ? args.nodeRootVolumeSize : undefined;
            resourceInputs["nodeRootVolumeThroughput"] = args ? args.nodeRootVolumeThroughput : undefined;
            resourceInputs["nodeRootVolumeType"] = args ? args.nodeRootVolumeType : undefined;
            resourceInputs["nodeSecurityGroup"] = args ? args.nodeSecurityGroup : undefined;
            resourceInputs["nodeSecurityGroupId"] = args ? args.nodeSecurityGroupId : undefined;
            resourceInputs["nodeSubnetIds"] = args ? args.nodeSubnetIds : undefined;
            resourceInputs["nodeUserData"] = args ? args.nodeUserData : undefined;
            resourceInputs["nodeUserDataOverride"] = args ? args.nodeUserDataOverride : undefined;
            resourceInputs["nodeadmExtraOptions"] = args ? args.nodeadmExtraOptions : undefined;
            resourceInputs["operatingSystem"] = args ? args.operatingSystem : undefined;
            resourceInputs["spotPrice"] = args ? args.spotPrice : undefined;
            resourceInputs["taints"] = args ? args.taints : undefined;
            resourceInputs["version"] = args ? args.version : undefined;
            resourceInputs["autoScalingGroupName"] = undefined /*out*/;
            resourceInputs["cfnStack"] = undefined /*out*/;
        }
        else {
            resourceInputs["autoScalingGroupName"] = undefined /*out*/;
            resourceInputs["cfnStack"] = undefined /*out*/;
            resourceInputs["extraNodeSecurityGroups"] = undefined /*out*/;
            resourceInputs["nodeSecurityGroup"] = undefined /*out*/;
            resourceInputs["nodeSecurityGroupId"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(NodeGroup.__pulumiType, name, resourceInputs, opts, true /*remote*/);
    }
}
exports.NodeGroup = NodeGroup;
/** @internal */
NodeGroup.__pulumiType = 'eks:index:NodeGroup';
