// @pulumi/eks/vpcCniAddon.js
"use strict";
// *** WARNING: this file was generated by pulumi-gen-eks. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcCniAddon = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * VpcCniAddon manages the configuration of the Amazon VPC CNI plugin for Kubernetes by leveraging the EKS managed add-on.
 * For more information see: https://docs.aws.amazon.com/eks/latest/userguide/eks-add-ons.html
 */
class VpcCniAddon extends pulumi.ComponentResource {
    /**
     * Returns true if the given object is an instance of VpcCniAddon.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === VpcCniAddon.__pulumiType;
    }
    /**
     * Create a VpcCniAddon resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        var _a, _b;
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.clusterName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'clusterName'");
            }
            resourceInputs["addonVersion"] = args ? args.addonVersion : undefined;
            resourceInputs["clusterName"] = args ? args.clusterName : undefined;
            resourceInputs["clusterVersion"] = args ? args.clusterVersion : undefined;
            resourceInputs["cniConfigureRpfilter"] = args ? args.cniConfigureRpfilter : undefined;
            resourceInputs["cniCustomNetworkCfg"] = args ? args.cniCustomNetworkCfg : undefined;
            resourceInputs["cniExternalSnat"] = args ? args.cniExternalSnat : undefined;
            resourceInputs["configurationValues"] = args ? args.configurationValues : undefined;
            resourceInputs["customNetworkConfig"] = args ? args.customNetworkConfig : undefined;
            resourceInputs["disableTcpEarlyDemux"] = args ? args.disableTcpEarlyDemux : undefined;
            resourceInputs["enableNetworkPolicy"] = args ? args.enableNetworkPolicy : undefined;
            resourceInputs["enablePodEni"] = args ? args.enablePodEni : undefined;
            resourceInputs["enablePrefixDelegation"] = args ? args.enablePrefixDelegation : undefined;
            resourceInputs["eniConfigLabelDef"] = args ? args.eniConfigLabelDef : undefined;
            resourceInputs["eniMtu"] = args ? args.eniMtu : undefined;
            resourceInputs["externalSnat"] = args ? args.externalSnat : undefined;
            resourceInputs["logFile"] = args ? args.logFile : undefined;
            resourceInputs["logLevel"] = args ? args.logLevel : undefined;
            resourceInputs["nodePortSupport"] = args ? args.nodePortSupport : undefined;
            resourceInputs["resolveConflictsOnCreate"] = (_a = (args ? args.resolveConflictsOnCreate : undefined)) !== null && _a !== void 0 ? _a : "OVERWRITE";
            resourceInputs["resolveConflictsOnUpdate"] = (_b = (args ? args.resolveConflictsOnUpdate : undefined)) !== null && _b !== void 0 ? _b : "OVERWRITE";
            resourceInputs["securityContextPrivileged"] = args ? args.securityContextPrivileged : undefined;
            resourceInputs["serviceAccountRoleArn"] = args ? args.serviceAccountRoleArn : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["vethPrefix"] = args ? args.vethPrefix : undefined;
            resourceInputs["warmEniTarget"] = args ? args.warmEniTarget : undefined;
            resourceInputs["warmIpTarget"] = args ? args.warmIpTarget : undefined;
            resourceInputs["warmPrefixTarget"] = args ? args.warmPrefixTarget : undefined;
        }
        else {
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "eks:index:VpcCni" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(VpcCniAddon.__pulumiType, name, resourceInputs, opts, true /*remote*/);
    }
}
exports.VpcCniAddon = VpcCniAddon;
/** @internal */
VpcCniAddon.__pulumiType = 'eks:index:VpcCniAddon';
