// @pulumi/aws-native/customerprofiles/eventTrigger.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventTrigger = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * An event trigger resource of Amazon Connect Customer Profiles
 */
class EventTrigger extends pulumi.CustomResource {
    /**
     * Get an existing EventTrigger resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new EventTrigger(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of EventTrigger.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === EventTrigger.__pulumiType;
    }
    /**
     * Create a EventTrigger resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.domainName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'domainName'");
            }
            if ((!args || args.eventTriggerConditions === undefined) && !opts.urn) {
                throw new Error("Missing required property 'eventTriggerConditions'");
            }
            if ((!args || args.objectTypeName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'objectTypeName'");
            }
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["domainName"] = args ? args.domainName : undefined;
            resourceInputs["eventTriggerConditions"] = args ? args.eventTriggerConditions : undefined;
            resourceInputs["eventTriggerLimits"] = args ? args.eventTriggerLimits : undefined;
            resourceInputs["eventTriggerName"] = args ? args.eventTriggerName : undefined;
            resourceInputs["objectTypeName"] = args ? args.objectTypeName : undefined;
            resourceInputs["segmentFilter"] = args ? args.segmentFilter : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["lastUpdatedAt"] = undefined /*out*/;
        }
        else {
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["description"] = undefined /*out*/;
            resourceInputs["domainName"] = undefined /*out*/;
            resourceInputs["eventTriggerConditions"] = undefined /*out*/;
            resourceInputs["eventTriggerLimits"] = undefined /*out*/;
            resourceInputs["eventTriggerName"] = undefined /*out*/;
            resourceInputs["lastUpdatedAt"] = undefined /*out*/;
            resourceInputs["objectTypeName"] = undefined /*out*/;
            resourceInputs["segmentFilter"] = undefined /*out*/;
            resourceInputs["tags"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const replaceOnChanges = { replaceOnChanges: ["domainName", "eventTriggerName"] };
        opts = pulumi.mergeOptions(opts, replaceOnChanges);
        super(EventTrigger.__pulumiType, name, resourceInputs, opts);
    }
}
exports.EventTrigger = EventTrigger;
/** @internal */
EventTrigger.__pulumiType = 'aws-native:customerprofiles:EventTrigger';
