// @pulumi/aws-compliance-policies/aws/eks/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cluster = void 0;
const eks_1 = require("@pulumi/aws/eks");
const policy_1 = require("@pulumi/policy");
const compliance_policy_manager_1 = require("@pulumi/compliance-policy-manager");
var Cluster;
(function (Cluster) {
    /**
     * Check that EKS Clusters API Endpoint are not publicly accessible.
     *
     * @severity critical
     * @frameworks hitrust, iso27001, pcidss
     * @topics network
     * @link https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html
     */
    Cluster.disallowApiEndpointPublicAccess = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-eks-cluster-disallow-api-endpoint-public-access",
            description: "Check that EKS Clusters API Endpoint are not publicly accessible.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(eks_1.Cluster, (cluster, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (cluster.vpcConfig.endpointPublicAccess === undefined || cluster.vpcConfig.endpointPublicAccess === true) {
                    /**
                     * We need to check `publicAccessCidrs` for any `0.0.0.0/0`.
                     */
                    if (!cluster.vpcConfig.publicAccessCidrs) {
                        reportViolation("EKS Cluster Endpoint API should not be publicly accessible.");
                    }
                    else {
                        if (cluster.vpcConfig.publicAccessCidrs.includes("0.0.0.0/0") || cluster.vpcConfig.publicAccessCidrs.includes("::/0")) {
                            reportViolation("EKS Cluster Endpoint API should not be publicly accessible.");
                        }
                    }
                }
            }),
        },
        vendors: ["aws"],
        services: ["eks"],
        severity: "critical",
        topics: ["network"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
    /**
     * Check that EKS Cluster Encryption Config is enabled.
     *
     * @severity high
     * @frameworks hitrust, iso27001, pcidss
     * @topics encryption, kubernetes
     * @link https://aws.amazon.com/blogs/containers/using-eks-encryption-provider-support-for-defense-in-depth/
     */
    Cluster.enableClusterEncryptionConfig = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-eks-cluster-enable-cluster-encryption-config",
            description: "Check that EKS Cluster Encryption Config is enabled.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(eks_1.Cluster, (cluster, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!cluster.encryptionConfig) {
                    reportViolation("EKS Cluster Encryption Configuration should be enabled.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["eks"],
        severity: "high",
        topics: ["encryption", "kubernetes"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
})(Cluster || (Cluster = {}));
exports.Cluster = Cluster;
