// @pulumi/aws-compliance-policies/aws/lambda/index.js
"use strict";
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ------------------------------- WARNING -------------------------------------
// This file was programmatically generated. Do not edit unless you know what
// you're doing.
// ------------------------------- WARNING -------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.Permission = exports.Function = void 0;
const lambda_1 = require("@pulumi/aws/lambda");
const policy_1 = require("@pulumi/policy");
const compliance_policy_manager_1 = require("@pulumi/compliance-policy-manager");
const lambda_2 = require("@pulumi/aws/lambda");
var Function;
(function (Function) {
    /**
     * Checks that Lambda functions have tracing configured.
     *
     * @severity low
     * @frameworks none
     * @topics logging, performance
     * @link https://docs.aws.amazon.com/lambda/latest/dg/services-xray.html
     */
    Function.configureTracingConfig = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-lambda-function-configure-tracing-config",
            description: "Checks that Lambda functions have tracing configured.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(lambda_1.Function, (f, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (f.tracingConfig && f.tracingConfig.mode !== "Active") {
                    reportViolation("Lambda functions should have tracing configured.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["lambda"],
        severity: "low",
        topics: ["logging", "performance"],
    });
    /**
     * Checks that Lambda functions have tracing enabled.
     *
     * @severity low
     * @frameworks none
     * @topics logging, performance
     * @link https://docs.aws.amazon.com/lambda/latest/dg/services-xray.html
     */
    Function.enableTracingConfig = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-lambda-function-enable-tracing-config",
            description: "Checks that Lambda functions have tracing enabled.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(lambda_1.Function, (f, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!f.tracingConfig) {
                    reportViolation("Lambda functions should have tracing enabled.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["lambda"],
        severity: "low",
        topics: ["logging", "performance"],
    });
    /**
     * Checks that all Lambda Functions have a description.
     *
     * @severity low
     * @frameworks none
     * @topics documentation
     * @link https://docs.aws.amazon.com/lambda/latest/dg/getting-started.html
     */
    Function.missingDescription = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-lambda-function-missing-description",
            description: "Checks that all Lambda Functions have a description.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(lambda_1.Function, (f, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!f.description) {
                    reportViolation("Lambda functions should have a description.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["lambda"],
        severity: "low",
        topics: ["documentation"],
    });
})(Function || (Function = {}));
exports.Function = Function;
var Permission;
(function (Permission) {
    /**
     * Checks that lambda function permissions have a source arn specified.
     *
     * @severity critical
     * @frameworks hitrust, iso27001, pcidss
     * @topics permissions, security
     * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-permission.html
     */
    Permission.configureSourceArn = compliance_policy_manager_1.policyManager.registerPolicy({
        resourceValidationPolicy: {
            name: "aws-lambda-permission-configure-source-arn",
            description: "Checks that lambda function permissions have a source arn specified.",
            configSchema: compliance_policy_manager_1.policyManager.policyConfigSchema,
            enforcementLevel: "advisory",
            validateResource: (0, policy_1.validateResourceOfType)(lambda_2.Permission, (f, args, reportViolation) => {
                if (!compliance_policy_manager_1.policyManager.shouldEvalPolicy(args)) {
                    return;
                }
                if (!f.sourceArn) {
                    reportViolation("Lambda functions permissions should have a source ARN defined.");
                }
            }),
        },
        vendors: ["aws"],
        services: ["lambda"],
        severity: "critical",
        topics: ["permissions", "security"],
        frameworks: ["pcidss", "hitrust", "iso27001"],
    });
})(Permission || (Permission = {}));
exports.Permission = Permission;
